rust/src/transformations/make_stable_expr/expr_binary/mod.rs

```diff
@@ -0,0 +1,78 @@
+use polars::prelude::*;
+use polars_plan::dsl::Expr;
+use polars_plan::utils::expr_output_name;
+
+use crate::core::{Function, MetricSpace, StabilityMap, Transformation};
+use crate::domains::{
+    AtomDomain, ExprContext, ExprDomain, LogicalPlanDomain, OuterMetric, SeriesDomain,
+};
+use crate::error::*;
+use crate::transformations::DatasetMetric;
+
+use super::StableExpr;
+
+#[cfg(test)]
+mod test;
+
+/// Make a Transformation that returns a binary expression
+///
+/// # Arguments
+/// * `input_domain` - Expr domain
+/// * `input_metric` - The metric under which neighboring LazyFrames are compared
+/// * `expr` - The clipping expression
+pub fn make_expr_binary<M: OuterMetric>(
+    input_domain: ExprDomain,
+    input_metric: M,
+    expr: Expr,
+) -> Fallible<Transformation<ExprDomain, ExprDomain, M, M>>
+where
+    M::InnerMetric: DatasetMetric,
+    M::Distance: Clone,
+    (ExprDomain, M): MetricSpace,
+    Expr: StableExpr<M, M>,
+{
+    let Expr::BinaryExpr { left, op, right } = expr.clone() else {
+        return fallible!(MakeTransformation, "expected binary expression");
+    };
+
+    let ExprDomain {
+        frame_domain,
+        context,
+    } = input_domain.clone();
+
+    let expr_domain = ExprDomain::new(frame_domain, ExprContext::RowByRow);
+    let t_left = left.make_stable(expr_domain.clone(), input_metric.clone())?;
+    let t_right = right.make_stable(expr_domain.clone(), input_metric.clone())?;
+
+    use polars_plan::dsl::Operator::*;
+    if !matches!(op, Eq | NotEq | Lt | LtEq | Gt | GtEq | And | Or | Xor) {
+        return fallible!(MakeTransformation, "unsupported operator: {:?}. Only binary operations that emit booleans are currently supported.", op);
+    }
+
+    let mut series_domain =
+        SeriesDomain::new(&*expr_output_name(&expr)?, AtomDomain::<bool>::default());
+
+    series_domain.nullable = t_left.output_domain.active_series()?.nullable
+        || t_right.output_domain.active_series()?.nullable;
```

Comment:
Does polars define the semantics of these operations? I'm wondering if for the boolean operations `&&` might be appropriate here. For example I'd expect `TRUE || NULL == TRUE` rather than `TRUE || NULL == NULL`, but maybe that's not the way it works.

Reply:
Based on the tests, `and` and `xor` both propagate nulls. Therefore all operations are sound, and all operations except for `or` are complete. I added conditions here to make it complete as well.

Reply:
After more testing, lhs nullity or falsey can still propagate null. I'm just leaving it simple, if either side can be null, the output can be null.